Feat: Created German Culture Decision for the region

John-Timothy opened the pull request.

John-Timothy committed:
Feat: Created German Culture Decision for the region

HansKalsi commented on common/decisions/00_major_decisions_german_culture.txt:
```diff
@@ -0,0 +1,70 @@
+				culture = culture:polish
+				culture = culture:carantanian
+				culture = culture:hungarian
+				culture = culture:croatian
```
From Czech to Croatian, these are not germanic cultures - these seven are West Slavic (if you hover over the culture name on any character in game, you should see the culture group they belong to)

Here is a link to the Ck3 Wiki, it should take you to the right section, but in case not, search for 'Central Germanic':
 https://ck3.paradoxwikis.com/Culture#:~:text=Original-,Central%20Germanic,-Bavarian
 
 Those are all the ones we want here

HansKalsi commented on common/culture/cultures/00_unified_german.txt:
```diff
@@ -0,0 +1,29 @@
+﻿unified_german = {
+	color = { 0.5 0.5 0.5 }
+	created = 843.1.1
+	parents = { old_saxon }
```
This won't need a parent as it'll be original

HansKalsi commented on common/culture/cultures/00_unified_german.txt:
```diff
@@ -0,0 +1,29 @@
+﻿unified_german = {
+	color = { 0.5 0.5 0.5 }
+	created = 843.1.1
```
From what I can tell, created isn't needed, so remove this line if that is correct

HansKalsi commented on common/culture/cultures/00_unified_german.txt:
```diff
@@ -0,0 +1,29 @@
+﻿unified_german = {
+	color = { 0.5 0.5 0.5 }
```
I'd recommend `0, 0.2, 0.33` (Prussian Blue, hex code: 0, 49, 83)

HansKalsi commented on common/culture/cultures/00_unified_german.txt:
```diff
@@ -0,0 +1,29 @@
+		15 = caucasian_ginger
+		35 = caucasian_brown_hair
+		25 = caucasian_dark_hair
+	}
```
I'd suggest altering these integers to the following: `40`, `10 `,`30`,`20` respectively

HansKalsi commented on common/decisions/00_major_decisions_german_culture.txt:
```diff
@@ -0,0 +1,70 @@
+	is_valid = {
+		AND = {
+			completely_controls_region = german_unified_lands
+		}
```
Don't need the AND wrapper if it's just one `completely_controls_region`

HansKalsi commented on common/decisions/00_major_decisions_german_culture.txt:
```diff
@@ -0,0 +1,70 @@
+
+		}
+
+		convert_family_culture_and_notify_vassals_effect = {
```
Instead of this, you'll likely want to do this as it also converts vassals and their captials if they meet certain requirements:
```
# Convert every vassal and their capital to German
every_vassal = {
   # Only apply to ai
   limit = {
       is_ai = yes
   }
   set_culture_same_as = root
   capital_county = {
	set_county_culture = root.culture
   }
}
```

HansKalsi replied:
(of course before this triggers, you need to of set the root characters culture to German already for it to work (and before that to have done the innovation transfer logic))

To clarify:
- Update innovations
- Convert root character to German
- Convert all vassals and some of their lands to German

HansKalsi replied:
After some further thought, you'll probably want BOTH the above comment code and the code you've already code as you want to update your family aswell as vassals

HansKalsi commented on common/culture/cultures/00_unified_german.txt:
```diff
@@ -0,0 +1,29 @@
+﻿unified_german = {
```
<img width="485" alt="Screenshot 2024-11-14 at 14 38 58" src="https://github.com/user-attachments/assets/394ad69b-5d01-407b-aa38-97d5788d88f8">

FYI I found a german base game culture, which appears to align with a lot of what you've put in this file.

<img width="538" alt="Screenshot 2024-11-14 at 14 41 08" src="https://github.com/user-attachments/assets/de15bdf2-2b3d-442e-b65d-fa60dfd69e48">

Pay careful attention to the task in hand, as stated there, if a German one existed it should be used outright, rather than creating a new one as then we don't need to worry about any localization stuff

HansKalsi replied:
At this point we have a custom file, so keep it.
But bare this in mind for the future, as we want to make our development easier and fast, so reusing what we can is important

HansKalsi commented on localization/english/from_castles_to_bastilles_l_english.yml:
```diff
@@ -111,4 +111,12 @@
+  form_german_culture_decision: "Form Unified German Culture"
+  form_german_culture_decision_desc: "The time has come for us to unify our cultures into one"
+  form_german_culture_decision_tooltip: "Consolidate all Germanic people into one German culture"
+  form_german_culture_decision_effect_tooltip: "All Germanic people assimilate under one Unified German Culture"
```
Should be "...assimilate **_into_** one..."

HansKalsi commented on map_data/geographical_regions/custom_german_regions.txt:
```diff
@@ -0,0 +1,90 @@
+world_india_rajastan_test = {
```
I know this is still WiP, but you shouldn't need this region in the german_region file

HansKalsi replied:
You didn't remove `world_india_rajastan_test` from this file?

HansKalsi commented on map_data/geographical_regions/custom_german_regions.txt:
```diff
@@ -0,0 +1,90 @@
+		d_croatia
+		d_dalmatia
+		d_slavonia
+		d_ragusa
```
Croatia is not historical german lands

HansKalsi commented on map_data/geographical_regions/custom_german_regions.txt:
```diff
@@ -0,0 +1,90 @@
+		d_ungvar
+		d_visegrad
+		d_bihar
+		d_transylvania
```
Hungary is not historical german lands

HansKalsi commented on map_data/geographical_regions/custom_german_regions.txt:
```diff
@@ -0,0 +1,90 @@
+		d_halych
+		d_lesser_poland
+		d_volhynia
+		d_cherven_cities
```
Galicia-Volhynia is not historical german lands

HansKalsi commented on map_data/geographical_regions/custom_german_regions.txt:
```diff
@@ -0,0 +1,90 @@
+		d_upper_silesia
+		d_masuria
+		d_pomerelia
+		d_prussia
```
I'd accept `d_prussia`, you have to check ingame for any custom culture areas via the culture map mode.
But think a lot of these don't apply, such as silesia

HansKalsi commented on map_data/geographical_regions/custom_german_regions.txt:
```diff
@@ -0,0 +1,90 @@
+
+		#bohemia
+		d_bohemia
+		d_moravia
```
Bohemia is not historical german lands (it's czech)

HansKalsi commented on localization/english/from_castles_to_bastilles_l_english.yml:
```diff
@@ -111,4 +111,12 @@
+  form_german_culture_decision_tooltip: "Consolidate all Germanic people into one German culture"
+  form_german_culture_decision_effect_tooltip: "All Germanic people assimilate under one Unified German Culture"
+  form_german_culture_decision_confirm: "We are One!"
+  unified_german: "Unified German Culture"
```
`unified_german: "German"` is what you need, we don't want to be verbose

HansKalsi commented on common/decisions/00_major_decisions_german_culture.txt:
```diff
@@ -0,0 +1,70 @@
+			NEW_CULTURE = culture:unified_german
+		}
+
+		discover_all_eras = all
```
`discover_all_eras` looks incorrect here, as I would assume that unlocks everything.

Here's some code I found which allowed updating the culture innovations:
```
every_culture_global = {
    limit = { has_cultural_pillar = heritage_latin }

    add_innovation_if_missing_effect = { INNOVATION = innovation_men_at_arms }
    add_innovation_if_missing_effect = { INNOVATION = innovation_pike_columns }
    add_innovation_if_missing_effect = { INNOVATION = innovation_guilds }
}
```

Obviously, instead of `limit = { has_cultural_pillar = heritage_latin }` you'll want `limit = { culture = unified_german }`

Hence in theory you'd need something like:
```
every_culture_global = {
   # We only care about the culture the root character currently is
   limit = {
      culture = root.culture
   }
   save_scope_as = original_culture

   # Update all innovations for German to at least have those of the enactor's culture
   every_culture_group = {
      limit = {
         culture = unified_german
      }

     if = {
           limit = {
                scope:original_culture = {
                        has_innovation = innovation_quilted_armor
                }
            }
           add_innovation_if_missing_effect = { INNOVATION = innovation_quilted_armor }
    }
   # ... repeated for every innovation, to ensure it works test with a handful first
   }
}
```

John-Timothy replied:
Regarding this, I saw a code that will add innovations from culture to culture which is: 

			root.culture = {
				get_all_innovations_from = culture:unified_german
			}

HansKalsi replied:
Great find!
I looked up that in the base game codebase and saw it's usage is like so:

<img width="358" alt="Screenshot 2024-11-14 at 15 50 28" src="https://github.com/user-attachments/assets/3008b388-0c19-4656-8250-aba85aa2918e">

Hence I've added another comment with a suggested re-write for this bit of code

John-Timothy committed:
Refactor: Changes based on suggestions for Code Review; Also added Get_all_innovations_from to add culture-specific innovations to current innovation

John-Timothy commented:
Pushed new Changes based on the code review suggestions. Also added the culture innovation integration for two cultures:
i.e.:
			root.culture = {
				get_all_innovations_from = culture:unified_german
			}

HansKalsi commented on common/decisions/00_major_decisions_german_culture.txt:
```diff
@@ -48,16 +48,16 @@ form_german_culture_decision = {
-
+			root.culture = {
+				get_all_innovations_from = culture:unified_german
+			}
```
This should be:
```
culture:unified_german = {
	get_all_innovations_from = root.culture
}
```

HansKalsi commented on localization/english/from_castles_to_bastilles_l_english.yml:
```diff
@@ -111,4 +111,14 @@
+  form_german_culture_decision_confirm: "We are One!"
+  unified_german: "Unified German Culture"
+  german_unified_lands: "Unified German Land"
+  get_all_innovations_from: "Get All Innovations From"
```
Why add localization for `get_all_innovations_from`?
Localizations should only be added if we are expecting this to appear to the player

John-Timothy replied:
Removed that part in the localizations

John-Timothy committed:
Refactor: Adjustments based on PR suggestions

John-Timothy committed:
Refactor: Unified German Culture to German

John-Timothy committed:
Refactor: Applied root.culture to Get All Innovations from:

John-Timothy commented:
New commits for refactored scripts to resolve suggestions

HansKalsi commented on common/decisions/00_major_decisions_german_culture.txt:
```diff
@@ -0,0 +1,65 @@
+		hidden_effect = {
+			root.culture = {
+				get_all_innovations_from = root.culture
+			}
```
Very close, but be careful because the syntax matters:
`culture:unified_german` is needed as shown below as that's the culture you are applying this to.

```
culture:unified_german = {
	get_all_innovations_from = root.culture
}
```

HansKalsi replied:
This should be done FIRST, before any culture related alterations take effect

John-Timothy committed:
Refactor: Adjustment to Hidden Effect Culture

John-Timothy committed:
Refactor: Removed World India Rajastan Test for German Regions

John-Timothy committed:
Refactor: Added Every vassal Culture conversion

HansKalsi committed:
- Get the german culture unification working properly (only converting in the region required for the decision
- Added unified german culture history so it's in the correct era
- Added Kingdom of Rum to Ottomans control
- Fixed some minor localization issues (pre-existing)
- Added some additional detail to german unification decision text

HansKalsi committed:
- Re-added custom indian region for british raj stuff that was unexplainable removed...
- Added custom image for german unification decision
